soaktest/soaktest.go

```diff
@@ -0,0 +1,285 @@
+			if !ok {
+				err := fmt.Errorf("Failed to convert partition "+
+					"to float64 %v\n", ok)
+				return err
+			}
+			if len(parts) > 0 {
+				parts += ","
+			}
+			parts += fmt.Sprintf("%d", int(f))
+		}
+
+		rttValue := broker.(map[string]interface{})["rtt"]
+		avg, ok := rttValue.(map[string]interface{})["avg"].(float64)
+		if !ok {
+			err := fmt.Errorf("failed to convert avg to float64 %v", ok)
+			return err
+		}
+
+		p99, ok = rttValue.(map[string]interface{})["p99"].(float64)
+		if !ok {
+			err := fmt.Errorf("failed to convert p99 to float64 %v", ok)
+			return err
+		}
+
+		nid, ok := broker.(map[string]interface{})["nodeid"].(float64)
+		if !ok {
+			err := fmt.Errorf("failed to convert nodeid to float64 %v", ok)
+			return err
+		}
+		nodeid = int(nid)
+
+		tags := []string{fmt.Sprintf("broker:%d, type:%s, topic:%s, parts:%s",
+			nodeid, raw["type"], topic, parts)}
+
+		DatadogGauge(brokerRttP99, p99/1000000.0, tags)
+		DatadogGauge(brokerRttAvg, avg/1000000.0, tags)
+	}
+	return nil
+}
+
+// HandleStatsEvent converts Stats events to hashmap
+func HandleStatsEvent(e *kafka.Stats, topic string) error {
+	var raw map[string]interface{}
+	err := json.Unmarshal([]byte(e.String()), &raw)
+	if err != nil {
+		ErrorLogger.Printf("Json unmarshall error: %s\n", err)
+		return err
+	}
+
+	err = rttStats(raw, topic)
+	if err != nil {
+		ErrorLogger.Printf("Failed to calculate broker rtt statistics: %s",
+			err)
+		return err
+	}
+	return nil
+}
+
+// ProducerDeliveryCheck handles delivery report for producer
+func ProducerDeliveryCheck(e *kafka.Message) {
+	tags := []string{fmt.Sprintf("topic:%s, partition:%d",
+		*e.TopicPartition.Topic, e.TopicPartition.Partition)}
+	if e.TopicPartition.Error != nil {
+		drErrCnt++
+		DatadogIncrement(producerDrErr, 1, tags)
+		ErrorLogger.Printf("Delivery failed: %v\n", e.TopicPartition)
+	} else {
+		drCnt++
+		DatadogIncrement(producerDr, 1, tags)
+		if drCnt%1000 == 0 {
+			InfoLogger.Printf("Delivered message to topic at events channel "+
+				"%s [%d] at offset %d\n", *e.TopicPartition.Topic,
+				e.TopicPartition.Partition, e.TopicPartition.Offset)
+		}
+	}
+}
+
+// PrintConsumerStatus prints the information for the consumer
+func PrintConsumerStatus() {
+	InfoLogger.Printf("Consumer: %d messages consumed, %d duplicates, "+
+		"%d missed, %d message errors, %d consumer client-level errors\n",
+		consumerMsgCnt, msgDupCntForOutput, msgMissCntForOutput,
+		consumerMsgErrs, ConsumerErrCnt)
+}
+
+// PrintProducerStatus prints the information for the producer
+func PrintProducerStatus() {
+	InfoLogger.Printf("Producer: %d messages produced, %d delivered, %d "+
+		"failed to deliver, %d producer client-level errors\n",
+		ProduceMsgCnt, drCnt, drErrCnt, ProducerErrCnt)
+}
+
+// HandleMessage handles received messages, monitors latency and
+// verifies messages
+func HandleMessage(e *kafka.Message,
+	hwmarks map[string]uint64) bool {
+	tags := []string{fmt.Sprintf("topic:%s, partition:%d",
```

Comment:
Should topic: and partition: perhaps be separate entries in the tags list vector? It seems odd that it is multidimensional like this